compiler/src/dotty/tools/dotc/typer/Nullables.scala

```diff
@@ -0,0 +1,341 @@
+     */
+    def nullableContextIf(c: Boolean)(given Context): Context =
+      val info = tree.notNullInfoIf(c)
+      if info.isEmpty then curCtx else curCtx.addNotNullInfo(info)
+
+    /** The context to use for the arguments of the function represented by `tree`.
+     *  This is the current context, augmented with nullability information
+     *  of the left argument, if the application is a boolean `&&` or `||`.
+     */
+    def nullableInArgContext(given Context): Context = tree match
+      case Select(x, _) if !curCtx.erasedTypes =>
+        if tree.symbol == defn.Boolean_&& then x.nullableContextIf(true)
+        else if tree.symbol == defn.Boolean_|| then x.nullableContextIf(false)
+        else curCtx
+      case _ => curCtx
+
+    /** The `tree` augmented with nullability information in an attachment.
+     *  The following operations lead to nullability info being recorded:
+     *
+     *  1. Null tests using `==`, `!=`, `eq`, `ne`, if the compared entity is
+     *     a path (i.e. a stable TermRef)
+     *  2. Boolean &&, ||, !
+     */
+    def computeNullable()(given Context): tree.type =
+      def setConditional(ifTrue: Set[TermRef], ifFalse: Set[TermRef]) =
+        tree.putAttachment(NNConditional, NotNullConditional(ifTrue, ifFalse))
+      if !curCtx.erasedTypes && analyzedOps.contains(tree.symbol.name.toTermName) then
+        tree match
+          case CompareNull(TrackedRef(ref), testEqual) =>
+            if testEqual then setConditional(Set(), Set(ref))
+            else setConditional(Set(ref), Set())
+          case Apply(Select(x, _), y :: Nil) =>
+            val xc = x.notNullConditional
+            val yc = y.notNullConditional
+            if !(xc.isEmpty && yc.isEmpty) then
+              if tree.symbol == defn.Boolean_&& then
+                setConditional(xc.ifTrue | yc.ifTrue, xc.ifFalse & yc.ifFalse)
+              else if tree.symbol == defn.Boolean_|| then
+                setConditional(xc.ifTrue & yc.ifTrue, xc.ifFalse | yc.ifFalse)
+          case Select(x, _) if tree.symbol == defn.Boolean_! =>
+            val xc = x.notNullConditional
+            if !xc.isEmpty then
+              setConditional(xc.ifFalse, xc.ifTrue)
+          case _ =>
+      tree
+
+    /** Compute nullability information for this tree and all its subtrees */
+    def computeNullableDeeply()(given Context): Unit =
+      new TreeTraverser {
+        def traverse(tree: Tree)(implicit ctx: Context) =
+          traverseChildren(tree)
+          tree.computeNullable()
+      }.traverse(tree)
+
+  given (tree: Assign)
+    def computeAssignNullable()(given Context): tree.type = tree.lhs match
+      case TrackedRef(ref) =>
+        tree.withNotNullInfo(NotNullInfo(Set(), Set(ref))) // TODO: refine with nullability type info
+      case _ => tree
+
+  private val analyzedOps = Set(nme.EQ, nme.NE, nme.eq, nme.ne, nme.ZAND, nme.ZOR, nme.UNARY_!)
+
+  /** A map from (name-) offsets of all local variables in this compilation unit
```

Comment:
Consider adding to the comment that we track the local variables through their offset and not through their name because of shadowing.